### PR TITLE
Graph layout using OGDF

### PR DESCRIPTION
**Pull request by wavewave**
OGDF C++ library is integrated to calculate graph layout on the fly and extract coordinates for display on the web. The binding is generated by fficxx, so relevant dependencies are added. 
Maybe this is used in the prototype phase and replaced later if a better solution is found. (as rather heavy dependencies from this is a downside.)

### CODE REVIEW

File: app/toolbox-server/Toolbox/Render/ModuleGraph.hs [hunk 1]
```diff
@@ -112,3 +157,73 @@ renderModuleGraph ss =
...
+        canvasWidth :: Double <- realToFrac <$> dRect_width drect
+        canvasHeight :: Double <- realToFrac <$> dRect_height drect
+
+        nodeLayout0 <- getAllNodeLayout g ga
+        let nodeLayout = mapMaybe replace nodeLayout0
+              where
+                replace (j, x, y, w, h) = do
+                  i <- L.lookup j moduleNodeIndexRevList
+                  name <- L.lookup i nameMap
+                  pure (name, x, y, w, h)
```
Comment (parsonsmatt) [inline app/toolbox-server/Toolbox/Render/ModuleGraph.hs:224]: this does seem like it'd be better as `IM.lookup` - especially on `nameMap`, we're doing two loops and doing a `lookup` on the list for each loop.

**Review by parsonsmatt — CHANGES_REQUESTED** [review]
Overall looks neat!

I think there are some data structure choices that could be more efficient - mostly just translating a few `[(Int, a)]` into `IntMap a`.

I'd also like to see fewer tuples and more named records - especially the `GraphVisInfo` type

File: app/toolbox-server/Toolbox/Render/ModuleGraph.hs (hunk 1 above)
Comment (wavewave) [reply app/toolbox-server/Toolbox/Render/ModuleGraph.hs:224]: heh. yeah. I kept using lists at the moment for prototyping (as it didn't hit performance problem yet) and was postponing to replace that globally. I think now is a good time to do that action. Thanks for the change request!

File: app/toolbox-server/Toolbox/Render/ModuleGraph.hs [hunk 2]
```diff
@@ -112,3 +157,73 @@ renderModuleGraph ss =
...
+      seeds = filterOutSmallNodes mgi
+      tordVtxs = topSort g
+      tordSeeds = filter (`elem` seeds) tordVtxs
+      reducedGraph = makeReducedGraph g tordSeeds
+   in IM.toList $ mkRevDep reducedGraph
+
+layOutGraph :: [(Int, Text)] -> [(Int, [Int])] -> IO GraphVisInfo
+layOutGraph nameMap graph = do
+  bracket newGraph delete $ \g ->
+    bracket (newGA g) delete $ \ga -> do
```
[the remark this answers is not in the document]
Comment (wavewave) [reply app/toolbox-server/Toolbox/Render/ModuleGraph.hs:178]: yes, indeed such safer resource management is needed. Here i did a minimal bracket. That introduction of ResourceT monad stack (as GraphLayout monad or something) is worth to be a separate PR, so I will make a PR to this branch. Thanks!

File: src/Plugin/Timing.hs [hunk 3]
```diff
@@ -96,11 +102,7 @@ extractModuleGraphInfo modGraph = do
         mapMaybe
           (\v -> (G.node_key v,) <$> modNameFromVertex v)
           vtxs
-      modNameRevMap = fmap swap modNameMap
-      topSorted =
-        mapMaybe
-          (\n -> L.lookup n modNameRevMap)
-          $ getTopSortedModules modGraph
+      topSorted = []  -- FOR NOW!
```
[the remark this answers is not in the document]
Comment (wavewave) [reply src/Plugin/Timing.hs]: Ah, right! This was disabled due to message update race condition happening. It was exacerbated with this topSort somehow (maybe the message grew up) but now the race condition was handled by #14. I am now reverting this. Thanks!

File: app/toolbox-server/Toolbox/Render/ModuleGraph.hs (hunk 1 above)
Comment (wavewave) [reply app/toolbox-server/Toolbox/Render/ModuleGraph.hs:224]: Now converted to IntMap #20

File: app/toolbox-server/Toolbox/Render/ModuleGraph.hs [hunk 4]
```diff
@@ -112,3 +157,73 @@ renderModuleGraph ss =
...
+        IM.fromList . concat
+          <$> ( forM graph $ \(i, _) -> do
+                  case L.lookup i nameMap of
+                    Nothing -> pure []
+                    Just name -> do
+                      let width = 8 * T.length name
+                      node <- newGraphNodeWithSize (g, ga) (width, 15)
+                      appendText ga node name
+                      pure [(i, node)]
+              )
```
[the remark this answers is not in the document]
Comment (wavewave) [reply app/toolbox-server/Toolbox/Render/ModuleGraph.hs:189]: converted to IntMap in #20

File: app/toolbox-server/Toolbox/Render/ModuleGraph.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (wavewave) [reply app/toolbox-server/Toolbox/Render/ModuleGraph.hs:179]: created a PR for this: #21

File: app/toolbox-server/Toolbox/Render/ModuleGraph.hs [hunk 5]
```diff
@@ -112,3 +157,73 @@ renderModuleGraph ss =
...
+      es = makeEdges $ mginfoModuleDep mgi
+      g = buildG (1, nVtx) es
+      seeds = filterOutSmallNodes mgi
+      tordVtxs = topSort g
+      tordSeeds = filter (`elem` seeds) tordVtxs
+      reducedGraph = makeReducedGraph g tordSeeds
+   in IM.toList $ mkRevDep reducedGraph
+
+layOutGraph :: [(Int, Text)] -> [(Int, [Int])] -> IO GraphVisInfo
+layOutGraph nameMap graph = do
```
[the remark this answers is not in the document]
Comment (wavewave) [reply app/toolbox-server/Toolbox/Render/ModuleGraph.hs]: Now I use IntMap approach for graph :-)

File: app/toolbox-server/Toolbox/Server/Types.hs [hunk 6]
```diff
@@ -24,6 +25,16 @@ data UIState = UIState
...
   }
 
+data GraphVisInfo = GraphVisInfo
+  { gviCanvasDim :: (Double, Double)
+  -- ^ (width, height)
+  , gviNodes :: [(Text, Double, Double, Double, Double)]
+  -- ^ (module name, x, y, w, h)
+  , gviEdges :: [(Int, [(Double, Double)])]
+  -- ^ (edge id, bend points)
+  }
```
[the remark this answers is not in the document]
Comment (wavewave) [reply app/toolbox-server/Toolbox/Server/Types.hs]: made new data types, Dimension, Point, NodeLayout, EdgeLayout. Thanks for suggestion!

**Review by parsonsmatt — APPROVED** [review]
Looks fantastic, thank you 😄

**Pull request merged**